### PR TITLE
Cleanup + improve abstraction

### PR DESCRIPTION
**Pull request by jonyas**
I tried to add a bit more abstraction layer to make viewModels simple to handle. I also did a big cleanup on repository layer, etc.

I really like this approach, and seems to work pretty nicely!

Enjoy your vacations!

### CODE REVIEW

File: app/src/main/java/anstaendig/com/architecturecomponents/repository/RepositoryImpl.kt [hunk 1]
```diff
@@ -16,50 +11,39 @@ class RepositoryImpl
...
-  }
-
   private val loadPerson: ObservableTransformer<Action.LoadPerson, Result> = ObservableTransformer { actions ->
     actions
         .flatMap { (id) ->
           swapiService
               .loadPerson(id)
               .subscribeOn(Schedulers.io())
               .toObservable()
+              .map<Result> { personData -> Result.Success(personData) }
```
Comment (anstaendig) [inline app/src/main/java/anstaendig/com/architecturecomponents/repository/RepositoryImpl.kt]: Should that not be rather outside of the flatMap? What is the reason for moving it?

File: app/src/main/java/anstaendig/com/architecturecomponents/ui/MainActivity.kt [hunk 2]
```diff
@@ -5,41 +5,38 @@ import android.widget.EditText
...
+class MainActivity : BaseActivity<MainActivityViewModel, MainActivityViewState, MainActivityUiEvent>() {
 
   private val messageTextView: TextView by bindView(R.id.message)
   private val editTextView: EditText by bindView(R.id.edit)
 
-  private val changeTextObservable: Observable<UiEvent.OnTextChange> by lazy {
+  override val layoutResource = R.layout.activity_main
+  override val viewModelClass = MainActivityViewModel::class.java
+
+  private val changeTextObservable: Observable<MainActivityUiEvent.OnTextChange> by lazy {
```
Comment (anstaendig) [inline app/src/main/java/anstaendig/com/architecturecomponents/ui/MainActivity.kt]: lazy(LazyThreadSafetyMode.NONE) as we don't need expensive locking on the main thread.

File: app/src/main/java/anstaendig/com/architecturecomponents/ui/MainActivity.kt [hunk 3]
```diff
@@ -5,41 +5,38 @@ import android.widget.EditText
...
     editTextView.textChanges()
-        .filter { text -> text.isNotBlank() }
-        .map { UiEvent.OnTextChange(it.toString()) }
+        .filter { text -> text.isNotEmpty() }
+        .map { MainActivityUiEvent.OnTextChange(it.toString()) }
   }
 
-  private val messageClickObservable: Observable<UiEvent.OnMessageClick> by lazy {
+  private val messageClickObservable: Observable<MainActivityUiEvent.OnMessageClick> by lazy {
```
Comment (anstaendig) [inline app/src/main/java/anstaendig/com/architecturecomponents/ui/MainActivity.kt]: lazy(LazyThreadSafetyMode.NONE)

File: app/src/main/java/anstaendig/com/architecturecomponents/ui/MainActivity.kt [hunk 4]
```diff
@@ -5,41 +5,38 @@ import android.widget.EditText
...
+        .map { MainActivityUiEvent.OnMessageClick(messageTextView.text.toString()) }
   }
 
-  private val events: Observable<UiEvent> by lazy {
-    Observable.merge(changeTextObservable, messageClickObservable)
-  }
-
-  override val layoutResource = R.layout.activity_main
-  override val viewModelClass = MainActivityViewModel::class.java
+  override val events: Observable<MainActivityUiEvent> by lazy { Observable.merge(changeTextObservable, messageClickObservable) }
```
Comment (anstaendig) [inline app/src/main/java/anstaendig/com/architecturecomponents/ui/MainActivity.kt]: lazy(LazyThreadSafetyMode.NONE)

File: app/src/main/java/anstaendig/com/architecturecomponents/ui/base/BaseActivity.kt [hunk 5]
```diff
@@ -37,14 +40,17 @@ abstract class BaseActivity<M : BaseViewModel<S>, S : BaseViewState>
...
       state?.let { render(it) }
     })
+    disposables.addAll(events.subscribe { event ->
+      viewModel.events.onNext(event)
+    })
   }
 
-  override fun onPause() {
+  override fun onDestroy() {
     disposables.dispose()
```
Comment (anstaendig) [inline app/src/main/java/anstaendig/com/architecturecomponents/ui/base/BaseActivity.kt]: I recognised this has to be the other way around.

File: app/src/main/java/anstaendig/com/architecturecomponents/repository/RepositoryImpl.kt (hunk 1 above)
Comment (jonduran) [inline app/src/main/java/anstaendig/com/architecturecomponents/repository/RepositoryImpl.kt]: doing that way, no need for RxRelay. Actions never fail :-)

File: app/src/main/java/anstaendig/com/architecturecomponents/ui/base/BaseActivity.kt (hunk 5 above)
Comment (jonduran) [inline app/src/main/java/anstaendig/com/architecturecomponents/ui/base/BaseActivity.kt]: What do you mean? If doing it onPause -> If you lock the screen and coming back, you need to subscribe everything again.

**Comment by jonduran** [comment]
Fixed some of the comments.

**Comment by jonduran** [comment]
[Other-topic] I managed to have Room working with the example! Demo when you are back!

File: app/src/main/java/anstaendig/com/architecturecomponents/ui/base/BaseActivity.kt (hunk 5 above)
Comment (anstaendig) [inline app/src/main/java/anstaendig/com/architecturecomponents/ui/base/BaseActivity.kt]: I just meant that the super call has to be done first and then our own code, not the other way around. It's a mistake I made and I just recognised it.

**Comment by anstaendig** [comment]
Looks good! Looking forward to Room, see ya next week.

**Pull request merged**